File: src/solutions/scala/N3LongestSubstring.scala
package solutions.scala

import scala.annotation.tailrec

object N3LongestSubstring {

  def lengthOfLongestSubstring(s: String): Int = {
    @tailrec
    def lengthOfLongestSubstring(set: Set[Char], max: Int, i: Int, j: Int): Int =
      if (i < s.length && j < s.length) {
        if (!set.contains(s.charAt(j))) lengthOfLongestSubstring(set + s.charAt(j), Math.max(max, j + 1 - i), i, j + 1)
        else lengthOfLongestSubstring(set - s.charAt(i), max, i + 1, j)
      }
      else max

    lengthOfLongestSubstring(Set.empty, 0, 0, 0)
  }

  def lengthOfLongestSubstring2(s: String): Int = {
    @tailrec
    def lengthOfLongestSubstring(list: List[Char], current: List[Char], max: Int): Int = list match {
      case h :: t =>
        val index = current.indexOf(h)
        if (index == -1) lengthOfLongestSubstring(t, current :+ h, max)
        else lengthOfLongestSubstring(t, current.slice(index + 1, current.size) :+ h, Math.max(current.size, max))
      case Nil => Math.max(max, current.size)
    }

    lengthOfLongestSubstring(s.toList, List.empty, 0)
  }


  def main(args: Array[String]): Unit = {
    assert(lengthOfLongestSubstring("abcabcbb") == 3)
    assert(lengthOfLongestSubstring("bbbbb") == 1)
    assert(lengthOfLongestSubstring("pwwkew") == 3)
  }
}

File: src/solutions/java/N3LongestSubstring.java
package solutions.java;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class N3LongestSubstring {

    public static int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int max = 0;
        int i = 0;
        int j = 0;
        while (j < s.length()) {
            if (map.containsKey(s.charAt(j))) {
                i = Math.max(map.get(s.charAt(j)), i);
            }
            max = Math.max(max, j - i + 1);
            map.put(s.charAt(j), ++j);
        }
        return max;
    }

    public static int lengthOfLongestSubstring2(String s) {
        Set<Character> set = new HashSet<>();
        int max = 0;
        int i = 0;
        int j = 0;
        while (i < s.length() && j < s.length()) {
            if (!set.contains(s.charAt(j))) {
                set.add(s.charAt(j++));
                max = Math.max(max, j - i);
            }
            else {
                set.remove(s.charAt(i++));
            }
        }
        return max;
    }

    public static int lengthOfLongestSubstring3(String s) {
        int max = 0;
        for (int i = 0; i < s.length(); i++) {
            Set<Character> set = new HashSet<>();
            int curr = 0;
            for (int j = i; j < s.length() && !set.contains(s.charAt(j)); j++) {
                set.add(s.charAt(j));
                curr++;
            }
            max = Math.max(max, curr);
        }
        return max;
    }

    public static void main(String[] args) {
        assert lengthOfLongestSubstring("abcabcbb") == 3;
        assert lengthOfLongestSubstring("bbbbb") == 1;
        assert lengthOfLongestSubstring("pwwkew") == 3;
    }
}
